core/services/pipeline/graph.go

```diff
@@ -182,19 +95,95 @@ func (n *TaskDAGNode) Attributes() []encoding.Attribute {
 	return r
 }
 
-func (n *TaskDAGNode) outputs() []*TaskDAGNode {
-	var nodes []*TaskDAGNode
-	ns := n.g.From(n.ID())
-	for ns.Next() {
-		nodes = append(nodes, ns.Node().(*TaskDAGNode))
+//
+
+type Pipeline struct {
+	Tasks  map[int64]Task
+	tree   *Tree
+	Source string
+
+	// roots
+	// terminals / returns / outputs
+}
+
+func (p *Pipeline) UnmarshalText(bs []byte) (err error) {
+	parsed, err := Parse(bs)
+	*p = *parsed
+	return err
+}
+func (p *Pipeline) MinTimeout() (time.Duration, bool, error) {
+	var minTimeout time.Duration = 1<<63 - 1
+	var aTimeoutSet bool
+	for _, t := range p.Tasks {
+		if timeout, set := t.TaskTimeout(); set && timeout < minTimeout {
+			minTimeout = timeout
+			aTimeoutSet = true
+		}
 	}
-	return nodes
+	return minTimeout, aTimeoutSet, nil
 }
 
-func unwrapGraphNodes(nodes graph.Nodes) []*TaskDAGNode {
-	var out []*TaskDAGNode
-	for nodes.Next() {
-		out = append(out, nodes.Node().(*TaskDAGNode))
+func (p *Pipeline) TopoSort() []Task {
+	result := make([]Task, len(p.Tasks))
+
+	for id, task := range p.Tasks {
```

Comment:
This seems suspiciously dependent on the insert order, which AFAIK is not guaranteed to be the same as the task execution order. Does it matter?

Reply:
actually on second thought, since the id is a topo index couldn't we just make the Tasks a slice instead of a map to simplify here ?